### PR TITLE
Tokenized reactions - user tags and hash tags (Resolves #1077)

### LINKED ISSUE #1077: [SuperEditor] - Hashtags and @-mentions
**Update: The following is a re-write of this ticket so that it describes a desired feature**

Make it possible for clients to tokenize at-mentions and hash tags.

An at-mention is a tokenization that's triggered by an @ character, e.g., @matthew-carroll

A hashtag is a tokenization that's triggered by a #, e.g., #flutter 

This feature should include the ability to anchor popovers near the token, however this ticket is only for the tokenization behavior. See #1166 for popover anchors.

## ------ Original ticket kept for posterity --------
Hi, @matthew-carroll @angelosilvestre any idea or code snippet on rendering username mentions in the editor e.g @username, thanks.

### PR DESCRIPTION
**Pull request by matthew-carroll**
Tokenized reactions - stable tags, pattern tags, and action tags (Resolves #1077)

A "pattern tag" is identified based on a trigger and desired characters. It can be edited at any time.

A "stable tag" is composed and then submitted. Once submitted, it's not editable.

An "action tag" is composed and then submitted. Once submitted, the tag text is deleted, because the tag is intended to represent a desired action.

Each type of tag has been implemented with a new concept of a `SuperEditorPlugin`. A plugin makes it easy to bundle related behaviors, which would otherwise require the user to add all of those pieces individually, such as request handlers, reactions, keyboard handlers, etc.

These plugins can be added as follows:

```dart
final _userTagPlugin = StableTagPlugin();
final _hashTagPlugin = PatternTagPlugin();

Widget build(BuildContext context) {
  return SuperEditor(
    // ...
    plugins: {
      _userTagPlugin,
     _hashTagPlugin,
    },
  );
}
```

The `StableTagPlugin` provides access to a `StableTagIndex`, which you can use to lookup the tag that the user is currently composing, as well as all committed stable tags, and all cancelled stable tags in the document.

The `PatternTagPlugin` provides access to a `PatternTagIndex`, which you can use to lookup all the pattern tags in the document.

The Example app includes a new page that uses stable tags as user tags, a new page that uses pattern tags as hash tags, and a new page that uses action tags to convert text node types.

These plugins don't have to be used for these behaviors. Users can create their own version of these features using a similar approach.

### CODE REVIEW

**Comment by matthew-carroll** [comment]
CC @Jethro87 @jmatth @abidex4yemi

**Comment by matthew-carroll** [comment]
CC @transcendr

**Comment by matthew-carroll** [comment]
@angelosilvestre we're going to merge this in and continue work after Superlist is able to stress test it. Can you do a sanity check code review?

File: super_editor/lib/src/infrastructure/attribution_layout_bounds.dart [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+class _AttributionBounds {
+  const _AttributionBounds(this.attribution, this.rect);
+
+  final Attribution attribution;
+  final Rect rect;
+}
+
+/// Filter function that decides whether the text with the given [attribution]
+/// should have a widget boundary places around it.
+typedef AttributionBoundsSelector = bool Function(Attribution attribution);
```
Comment (angelosilvestre) [inline super_editor/lib/src/infrastructure/attribution_layout_bounds.dart:125]: places -> placed?

File: super_editor/test/super_editor/text_entry/pattern_tags_test.dart [hunk 2]
```diff
@@ -0,0 +1,463 @@
...
+        await _pumpTestEditor(
+          tester,
+          singleParagraphEmptyDoc(),
+        );
+        await tester.placeCaretInParagraph("1", 0);
+
+        // Insert a single "#".
+        await tester.typeImeText("#");
+
+        // Ensure that no hash tag was created
```
Comment (angelosilvestre) [inline super_editor/test/super_editor/text_entry/pattern_tags_test.dart:23]: Reminder to end the comment with a `.`

File: super_editor/test/super_editor/text_entry/pattern_tags_test.dart [hunk 3]
```diff
@@ -0,0 +1,463 @@
...
+        await tester.pressBackspace();
+        await tester.pressBackspace();
+
+        // Ensure that the tag has a composing attribution.
+        final text = SuperEditorInspector.findTextInParagraph("1");
+        expect(text.text, "#");
+        expect(
+          text.hasAttributionAt(0, attribution: const PatternTagAttribution()),
+          isFalse,
+        );
```
Comment (angelosilvestre) [inline super_editor/test/super_editor/text_entry/pattern_tags_test.dart:104]: "the tag has" -> "the tag doesn't have" ?

File: super_editor/test/super_editor/text_entry/pattern_tags_test.dart [hunk 4]
```diff
@@ -0,0 +1,463 @@
...
+
+        // Place the caret at "befor|e #flutter after"
+        await tester.placeCaretInParagraph("1", 5);
+
+        // Expand downstream until we push one character into the tag.
+        await tester.pressShiftRightArrow();
+        await tester.pressShiftRightArrow();
+        await tester.pressShiftRightArrow();
+
+        // Ensure that the extent pushed into the tag.
```
Comment (angelosilvestre) [inline super_editor/test/super_editor/text_entry/pattern_tags_test.dart:360]: "pushed" -> "was pushed" ?

File: super_editor/test/super_editor/text_entry/stable_tags_test.dart [hunk 5]
```diff
@@ -0,0 +1,1019 @@
...
+              ),
+            ],
+          ),
+        );
+
+        // Place the caret at "before |"
+        await tester.placeCaretInParagraph("1", 7);
+
+        // Composer a stable tag.
+        await tester.typeImeText("@john");
```
Comment (angelosilvestre) [inline super_editor/test/super_editor/text_entry/stable_tags_test.dart:152]: "Composer" -> "Composes"

File: super_editor/test/super_editor/text_entry/stable_tags_test.dart [hunk 6]
```diff
@@ -0,0 +1,1019 @@
...
+              ),
+            ],
+          ),
+        );
+
+        // Place the caret at "before | after"
+        await tester.placeCaretInParagraph("1", 7);
+
+        // Composer a stable tag.
+        await tester.typeImeText("@john");
```
Comment (angelosilvestre) [inline super_editor/test/super_editor/text_entry/stable_tags_test.dart:223]: "Composer" -> "Composes"

**Review by angelosilvestre — APPROVED** [review]
LGTM with a few minor comments.

File: super_editor/example/lib/demos/features/feature_hash_tags.dart
Comment (miguelcmedeiros) [inline super_editor/example/lib/demos/features/feature_hash_tags.dart:1]: Rename file to `feature_pattern_tags.dart`

File: super_editor/example/lib/demos/features/feature_user_tags.dart
Comment (miguelcmedeiros) [inline super_editor/example/lib/demos/features/feature_user_tags.dart:1]: Rename file to `feature_stable_tags.dart`

File: super_editor/lib/src/default_editor/text_tokenizing/action_tags.dart [hunk 7]
```diff
@@ -0,0 +1,488 @@
...
+    ]);
+
+    return ExecutionInstruction.haltExecution;
+  }
+}
+
+const defaultActionTagRule = TagRule(trigger: "/", excludedCharacters: {" "});
+
+class SubmitComposingActionTagRequest implements EditRequest {
+  // TODO:
```
Comment (miguelcmedeiros) [inline super_editor/lib/src/default_editor/text_tokenizing/action_tags.dart:112]: Please specify what's still left to do.

**Review by miguelcmedeiros — APPROVED** [review]

File: super_editor/lib/src/default_editor/text_tokenizing/stable_tags.dart [hunk 8]
```diff
@@ -0,0 +1,1535 @@
...
+}
+
+class FillInComposingUserTagCommand implements EditCommand {
+  const FillInComposingUserTagCommand(
+    this._tag,
+    this._tagRule,
+  );
+
+  final String _tag;
+  final TagRule _tagRule;
```
[the remark this answers is not in the document]
Comment (matthew-carroll) [reply super_editor/lib/src/default_editor/text_tokenizing/stable_tags.dart:172]: Yeah, I think so. Let me know if there's a reason that it shouldn't be private.

File: super_editor/lib/src/default_editor/text_tokenizing/stable_tags.dart [hunk 9]
```diff
@@ -0,0 +1,1535 @@
...
+          documentSelection: DocumentSelection(
+            base: cancelledToken.indexedTag.start,
+            extent: cancelledToken.indexedTag.end,
+          ),
+          attributions: {stableTagCancelledAttribution},
+        ),
+      ]);
+
+      return;
+    }
```
[the remark this answers is not in the document]
Comment (matthew-carroll) [reply super_editor/lib/src/default_editor/text_tokenizing/stable_tags.dart]: I think I reworked this behavior for other reasons and that this no longer applies. But if it moved somewhere else and you find it, let me know.

**Comment by matthew-carroll** [comment]
There are probably still bugs in this, and possibly some missing features. But we're going to merge this so that Superlist can try to use it. We'll move forward with whatever they find.

**Pull request merged**